### PR TITLE
Flatten inheritance for FSDP strategy

### PR DESCRIPTION
**Pull request by edward-io**
## What does this PR do?

DDPFullyShardedStrategy now inherits from ParallelStrategy instead of DDPStrategy.

Part of #[11863](https://github.com/PyTorchLightning/pytorch-lightning/issues/11863)

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by ananthsub** [comment]
@edward-io does this depend on https://github.com/PyTorchLightning/pytorch-lightning/pull/11643 ? Or should we merge this, and then use #11643 to remove the duplication?

fyi @rohitgr7 @awaelchli

File: pytorch_lightning/strategies/fully_sharded.py [hunk 1]
```diff
@@ -118,19 +136,66 @@ def __init__(
...
+        return self._num_nodes
+
+    @num_nodes.setter
+    def num_nodes(self, num_nodes: int) -> None:
+        # note that world ranks is related to num_nodes, when resetting it, need to reset world ranks
+        self._num_nodes = num_nodes
+        self.set_world_ranks()
+
+    @property
+    def num_processes(self):
```
Comment (ananthsub) [inline pytorch_lightning/strategies/fully_sharded.py:162]:
```suggestion
    def num_processes(self) -> int:
```

File: pytorch_lightning/strategies/fully_sharded.py [hunk 2]
```diff
@@ -145,6 +210,76 @@ def setup(self, trainer: "pl.Trainer") -> None:
         self.barrier()
         self.setup_optimizers(trainer)
 
+    def _check_can_spawn_children(self):
```
Comment (ananthsub) [inline pytorch_lightning/strategies/fully_sharded.py:213]:
```suggestion
    def _check_can_spawn_children(self) -> None:
```

File: pytorch_lightning/strategies/fully_sharded.py [hunk 3]
```diff
@@ -145,6 +210,76 @@ def setup(self, trainer: "pl.Trainer") -> None:
...
+        if self.local_rank != 0:
+            raise RuntimeError(
+                "Lightning attempted to launch new distributed processes with `local_rank > 0`. This should not happen."
+                " Possible reasons: 1) LOCAL_RANK environment variable was incorrectly modified by the user,"
+                " 2) `ClusterEnvironment.creates_processes_externally` incorrectly implemented."
+            )
+
+    def _call_children_scripts(self):
```
Comment (ananthsub) [inline pytorch_lightning/strategies/fully_sharded.py:221]:
```suggestion
    def _call_children_scripts(self) -> None:
```

File: pytorch_lightning/strategies/fully_sharded.py [hunk 4]
```diff
@@ -171,6 +306,42 @@ def wrap_policy(*args, **kwargs):
 
         log.detail(f"{self.__class__.__name__}: exiting model_sharded_context.")
 
+    def determine_ddp_device_ids(self):
```
Comment (ananthsub) [inline pytorch_lightning/strategies/fully_sharded.py:309]:
```suggestion
    def determine_ddp_device_ids(self) -> Optional[int]:
```

File: pytorch_lightning/strategies/fully_sharded.py [hunk 5]
```diff
@@ -171,6 +306,42 @@ def wrap_policy(*args, **kwargs):
...
+            torch.distributed.barrier(device_ids=self.determine_ddp_device_ids())
+        else:
+            torch.distributed.barrier()
+
+    def broadcast(self, obj: object, src: int = 0) -> object:
+        obj = [obj]
+        if self.global_rank != src:
+            obj = [None]
+        torch.distributed.broadcast_object_list(obj, src, group=_group.WORLD)
+        return obj[0]
```
Comment (ananthsub) [inline pytorch_lightning/strategies/fully_sharded.py:327]: https://github.com/PyTorchLightning/pytorch-lightning/blob/79c4e5de60685dbc895641b0139ffc6180d069aa/pytorch_lightning/strategies/strategy.py#L37

```suggestion
    def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
        obj = [obj]
        if self.global_rank != src:
            obj = [None]
        torch.distributed.broadcast_object_list(obj, src, group=_group.WORLD)
        return obj[0]
```

File: pytorch_lightning/strategies/fully_sharded.py (hunk 1 above)
Comment (four4fish) [reply pytorch_lightning/strategies/fully_sharded.py:162]: optional[int] i think

File: pytorch_lightning/strategies/fully_sharded.py (hunk 5 above)
Comment (four4fish) [reply pytorch_lightning/strategies/fully_sharded.py:327]: This is not necessarily true
- for pytorch version< 1.8, torch collective using broadcast, which returns Async work handle, if async_op is set to True. None, if not async_op or if not part of the group
- for pytorch version .=1.8 torch collective using broadcast, which returns None. If rank is part of the group, object_list will contain the broadcasted objects from src rank.

I think change obj to TBroadcast will cause issues

**Comment by edward-io** [comment]
> @edward-io does this depend on https://github.com/PyTorchLightning/pytorch-lightning/pull/11643 ? Or should we merge this, and then use https://github.com/PyTorchLightning/pytorch-lightning/pull/11643 to remove the duplication?

I believe we're doing the latter, @four4fish can confirm.

**Comment by edward-io** [comment]
Addressed typing comments.

**Comment by four4fish** [comment]
> @edward-io does this depend on https://github.com/PyTorchLightning/pytorch-lightning/pull/11643 ? Or should we merge this, and then use https://github.com/PyTorchLightning/pytorch-lightning/pull/11643 to remove the duplication?

I remember Adrian suggested merge 11643 first, I think it make sense to merge bigger PRs first too.

File: pytorch_lightning/strategies/fully_sharded.py [hunk 6]
```diff
@@ -183,12 +224,106 @@ def configure_ddp(self) -> None:
 
         # setup optimizers after fully sharded has wrapped the lightning module
         self.setup_optimizers(self.lightning_module.trainer)
+        self.barrier()
+        self.setup_optimizers(trainer)
```
Comment (four4fish) [inline pytorch_lightning/strategies/fully_sharded.py:228]: Here is duplicated.
Let's keep the previous ones, the setup_optimizer should happen before model_to_device

Comment (four4fish) [reply pytorch_lightning/strategies/fully_sharded.py:228]: @edward-io is exactly reproducing the logic here. 
@ananthsub @awaelchli  does move_to_device for FSDP happens before set optimizer as well? Was these part of the duplication necessary? 
The comments above mentioned that `FSDP module.to(device) would first summon all parameters`, but we are not using FSDP.apply or module.to. Am I missing something here?

Comment (edward-io) [reply pytorch_lightning/strategies/fully_sharded.py:228]: > Here is duplicated. Let's keep the previous ones, the setup_optimizer should happen before model_to_device

oops. sorry, this was a bad merge

File: pytorch_lightning/strategies/fully_sharded.py [hunk 7]
```diff
@@ -172,8 +205,16 @@ def wrap_policy(*args, **kwargs):
...
-    def configure_ddp(self) -> None:
-        log.detail(f"{self.__class__.__name__}: configuring DDP... (cpu_offload: [{self.cpu_offload}])")
+    def setup(self, trainer: "pl.Trainer") -> None:
+        self.accelerator.setup(trainer)
+        self.setup_optimizers(trainer)
+        self.setup_precision_plugin()
+        optimizers_to_device(self.optimizers, self.root_device)
+
+        if self.sync_batchnorm:
+            self.model = self.configure_sync_batchnorm(self.model)
```
Comment (ananthsub) [inline pytorch_lightning/strategies/fully_sharded.py:215]: note: might run into merge conflicts with #11754

File: pytorch_lightning/strategies/fully_sharded.py [hunk 8]
```diff
@@ -182,13 +223,106 @@ def configure_ddp(self) -> None:
...
+        # todo (tchaton) Add support for non-shared file-system which will fail.
+        if len(os.listdir(sync_dir)) == (self.world_size // self.num_nodes):
+            return
+
+        for pid in self._pids:
+            if pid != os.getpid():
+                os.kill(pid, signal.SIGKILL)
+        shutil.rmtree(sync_dir)
+        raise DeadlockDetectedException(f"DeadLock detected from rank: {self.global_rank} \n {trace}")
+
```
Comment (ananthsub) [inline pytorch_lightning/strategies/fully_sharded.py:325]: you can remove the override for `post_training_step` in L342 now

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/strategies/fully_sharded.py [hunk 9]
```diff
@@ -109,6 +121,8 @@ def __init__(
             checkpoint_io=checkpoint_io,
             precision_plugin=precision_plugin,
         )
+        self._num_nodes = 1
+        self.sync_batchnorm = False
```
Comment (awaelchli) [inline pytorch_lightning/strategies/fully_sharded.py:125]: needs rebase onto master (same in other prs) #11754

File: pytorch_lightning/strategies/fully_sharded.py [hunk 10]
```diff
@@ -205,9 +340,6 @@ def predict_step(self, *args, **kwargs) -> STEP_OUTPUT:
         with self.precision_plugin.predict_step_context():
             return self.model.predict_step(*args, **kwargs)
 
-    def post_training_step(self):
-        pass
-
```
Comment (awaelchli) [inline pytorch_lightning/strategies/fully_sharded.py:210]: pre_backward is missing too. it was implemented in ddp but the regular sharded strategy also ignores it. 
asking for confirmation that this is intentional

```py
    def pre_backward(self, closure_loss: torch.Tensor) -> None:
        """Run before precision plugin executes backward."""
        if not self.lightning_module.automatic_optimization:
            prepare_for_backward(self.model, closure_loss)
```

File: pytorch_lightning/strategies/fully_sharded.py [hunk 11]
```diff
@@ -205,9 +340,6 @@ def predict_step(self, *args, **kwargs) -> STEP_OUTPUT:
...
     @classmethod
     def register_strategies(cls, strategy_registry: Dict) -> None:
```
Comment (awaelchli) [inline pytorch_lightning/strategies/fully_sharded.py:344]: where is the teardown? I think we need it?

```py

    def teardown(self) -> None:
        log.detail(f"{self.__class__.__name__}: tearing down DDP plugin")
        super().teardown()
        if isinstance(self.model, DistributedDataParallel):
            self.model = self.lightning_module

        if self._layer_sync:
            self.model = self._layer_sync.revert(self.model)

        if self.root_device.type == "cuda":
            # GPU teardown
            log.detail(f"{self.__class__.__name__}: moving model to CPU")
            self.lightning_module.cpu()
            # clean up memory
            torch.cuda.empty_cache()

```

**Review by awaelchli — CHANGES_REQUESTED** [review]

File: pytorch_lightning/strategies/fully_sharded.py [hunk 12]
```diff
@@ -119,32 +132,53 @@ def __init__(
...
+
+    @property
+    def num_nodes(self) -> int:
+        return self._num_nodes
+
+    @num_nodes.setter
+    def num_nodes(self, num_nodes: int) -> None:
+        # note that world ranks is related to num_nodes, when resetting it, need to reset world ranks
+        self._num_nodes = num_nodes
+        self.set_world_ranks()
```
[the remark this answers is not in the document]
Comment (four4fish) [reply pytorch_lightning/strategies/fully_sharded.py:156]: I don't think this is needed anywhere. Accl_conn calls set_world_ranks() at the end, when every properties is final. https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/accelerator_connector.py#L739

Unless user do trainer.strategy.num_nodes=X, the num_nodes shouldn't change after trainer initialized. 
We shouldn't support runtime device change anyway, so I think we don't need self.set_world_ranks()

File: pytorch_lightning/strategies/fully_sharded.py [hunk 13]
```diff
@@ -119,32 +132,53 @@ def __init__(
...
+
+    @property
+    def num_processes(self) -> int:
```
[the remark this answers is not in the document]
Comment (four4fish) [reply pytorch_lightning/strategies/fully_sharded.py:159]: ParallelStrategy has distributed_sampler_kwargs method, I think len(self.parallel_devices) == num_processes*num_nodes?

File: pytorch_lightning/strategies/fully_sharded.py [hunk 14]
```diff
@@ -119,32 +132,53 @@ def __init__(
...
-    def setup(self, trainer: "pl.Trainer") -> None:
-        self.accelerator.setup(trainer)
-        self.setup_optimizers(trainer)
-        self.setup_precision_plugin()
-        optimizers_to_device(self.optimizers, self.root_device)
+        log.detail(f"{self.__class__.__name__}: setting up distributed...")
+        reset_seed()
 
         if self._layer_sync:
             self.model = self._layer_sync.apply(self.model)
```
[the remark this answers is not in the document]
Comment (four4fish) [reply pytorch_lightning/strategies/fully_sharded.py:172]: the sync batch norm was in setup() for both ddp and fsdp
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/strategies/fully_sharded.py#L142-L143

File: pytorch_lightning/strategies/fully_sharded.py [hunk 15]
```diff
@@ -119,32 +132,53 @@ def __init__(
         self.min_num_params = min_num_params
         self.state_dict_device = torch.device("cpu") if state_dict_to_cpu else None
         self._process_group = None
+        self._rank_0_will_call_children_scripts: bool = False
```
[the remark this answers is not in the document]
Comment (edward-io) [reply pytorch_lightning/strategies/fully_sharded.py:135]: > Where did _share_information_to_prevent_deadlock go?

That's right, since its not called in the setup method, its not needed in this file AFAICT.

File: pytorch_lightning/strategies/fully_sharded.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/strategies/fully_sharded.py:159]: `_is_single_process_single_device` is useful only for the DDP communication hooks, which don't take effect here. That's part of the motivation of flattening the inheritance is to see what's actually used.

File: pytorch_lightning/strategies/fully_sharded.py [hunk 16]
```diff
@@ -172,8 +210,16 @@ def wrap_policy(*args, **kwargs):
...
         log.detail(f"{self.__class__.__name__}: exiting model_sharded_context.")
 
-    def configure_ddp(self) -> None:
-        log.detail(f"{self.__class__.__name__}: configuring DDP... (cpu_offload: [{self.cpu_offload}])")
+    def setup(self, trainer: "pl.Trainer") -> None:
+        self.accelerator.setup(trainer)
+        self.setup_optimizers(trainer)
+        self.setup_precision_plugin()
+        optimizers_to_device(self.optimizers, self.root_device)
+
```
Comment (tchaton) [inline pytorch_lightning/strategies/fully_sharded.py:218]: Seems to be missing some logic from DDP there. `_rank_0_will_call_children_scripts`.

File: pytorch_lightning/strategies/fully_sharded.py [hunk 17]
```diff
@@ -172,8 +210,16 @@ def wrap_policy(*args, **kwargs):
 
         log.detail(f"{self.__class__.__name__}: exiting model_sharded_context.")
 
-    def configure_ddp(self) -> None:
-        log.detail(f"{self.__class__.__name__}: configuring DDP... (cpu_offload: [{self.cpu_offload}])")
+    def setup(self, trainer: "pl.Trainer") -> None:
+        self.accelerator.setup(trainer)
```
Comment (tchaton) [inline pytorch_lightning/strategies/fully_sharded.py:214]: should be super().setup(trainer) instead.

File: pytorch_lightning/strategies/fully_sharded.py [hunk 18]
```diff
@@ -172,8 +210,16 @@ def wrap_policy(*args, **kwargs):
...
+        self.setup_optimizers(trainer)
```
Comment (tchaton) [inline pytorch_lightning/strategies/fully_sharded.py:215]: Do you know why setup_optimizers is called 2 times?

File: pytorch_lightning/strategies/fully_sharded.py [hunk 19]
```diff
@@ -119,32 +132,57 @@ def __init__(
...
+    def num_nodes(self, num_nodes: int) -> None:
+        # note that world ranks is related to num_nodes, when resetting it, need to reset world ranks
+        self._num_nodes = num_nodes
+        self.set_world_ranks()
+
+    @property
+    def num_processes(self) -> int:
+        return len(self.parallel_devices) if self.parallel_devices is not None else 0
+
+    def setup_environment(self) -> None:
```
Comment (tchaton) [inline pytorch_lightning/strategies/fully_sharded.py:168]: Let's move this code into `setup_distributed` as DDP. Could maybe be moved to the Parallel class.

File: pytorch_lightning/strategies/fully_sharded.py [hunk 20]
```diff
@@ -182,13 +228,106 @@ def configure_ddp(self) -> None:
...
+            return
+        self.cluster_environment.set_global_rank(self.node_rank * self.num_processes + self.local_rank)
+        self.cluster_environment.set_world_size(self.num_nodes * self.num_processes)
+        rank_zero_only.rank = self.cluster_environment.global_rank()
+
+    def barrier(self, *args, **kwargs) -> None:
+        if not distributed_available():
+            return
+        if _TORCH_GREATER_EQUAL_1_8 and torch.distributed.get_backend() == "nccl":
+            if self.root_device.type == "cpu":
```
Comment (tchaton) [inline pytorch_lightning/strategies/fully_sharded.py:255]: Why it is different from master? Should be its own PR.

**Comment by Borda** [comment]
@awaelchli, is this still actual regarding the ongoing Lite refactor? :otter:
